diff --git a/bfd/elflink.c b/bfd/elflink.c
index 330fad79aa3..0d42c4c5f94 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -3851,6 +3851,7 @@ elf_link_add_object_symbols (bfd *abfd, struct bfd_link_info *info)
       if (!just_syms
 	  && (bfd_link_pic (info)
 	      || (!bfd_link_relocatable (info)
+		  && info->nointerp
 		  && (info->export_dynamic || info->dynamic)))
 	  && is_elf_hash_table (htab)
 	  && info->output_bfd->xvec == abfd->xvec
